### PR TITLE
ACAS-710: Code value picklists type ahead search

### PR DESCRIPTION
**Pull request by brianbolt**
## Description
 - See related acas-roo-server changes https://github.com/mcneilco/acas/pull/1148
 - Updated code table acas form widget to type ahead search rather than pre-fetch all picklist items.
 - In the case that there is a selected value, it will fetch the selected value only so that the intial value will be set. 

## Related Issue
ACAS-710

## How Has This Been Tested?
- Ran acasclient tests to verify compatability.
- Uploaded thousands of data column/ column tree ls_type / ls_kind values to ddict value and verified the code table fields behaved properly in Experiment editor and protocol editor as well as those in each of the endpoint manage fields.

Demo video
demo

### CODE REVIEW

File: modules/Components/src/client/ACASFormFields.coffee [hunk 1]
```diff
@@ -435,6 +439,9 @@ class ACASFormLSCodeValueFieldController extends ACASFormAbstractFieldController
 		if @options.autoSavePickListItem? && !@options.autoSavePickListItem
 			plOptions.autoSave = false
 
+		plOptions.autoFetch = true
+		if @options.autoFetch?
+			plOptions.autoFetch = @options.autoFetch
```
Comment (brianbolt) [inline modules/Components/src/client/ACASFormFields.coffee:444]: *sort of outside pr*: Previously there was no way to pass auto fetch to the picklist controller.  At one point I thought I could turn of auto fetching of the picklists but forgot about the "selectedCode" use case where the picklist actually needs to get populated with the selected value so it needs atleast that item int he picklist. 

TLDR; I left this in as a feature but it defaults to true like before.

File: modules/Components/src/client/PickList.coffee [hunk 2]
```diff
@@ -361,7 +361,28 @@ class PickListSelect2Controller extends PickListSelectController
 			openOnEnter: false
 			allowClear: true
 			width: @width
-
+			ajax:
```
Comment (brianbolt) [inline modules/Components/src/client/PickList.coffee:364]: This is most of the magic.  The select2 widget lets you add ajax calls for type ahead search.  I'm leveraging that.  I had to do this sort of hacky thing create a real URL to remove the "shortName" parameter if it existed in the list and only include the labelTextSearch and maxHits parameters.

File: modules/Components/src/client/PickList.coffee [hunk 3]
```diff
@@ -685,6 +706,7 @@ class EditablePickListSelect2Controller extends EditablePickListSelectController
 			collection: @collection
 			selectedCode: @options.selectedCode
 			filters: filters
+			autoFetch: @options.autoFetch
```
Comment (brianbolt) [inline modules/Components/src/client/PickList.coffee:709]: *sort of outside pr scope*: Pass options auto fetch to the picklist controller options as noted elsewhere in my review.

File: modules/Components/src/server/routes/CodeTableServiceRoutes.coffee [hunk 4]
```diff
@@ -37,31 +37,33 @@ exports.getAllCodeTableValues = (req, resp) ->
 		)
 
 exports.getCodeTableValues = (req, resp) ->
-	exports.getCodeTableValuesInternal req.params.type, req.params.kind, (result) ->
+	exports.getCodeTableValuesInternal req.params.type, req.params.kind, req.query, (result) ->
```
Comment (brianbolt) [inline modules/Components/src/server/routes/CodeTableServiceRoutes.coffee:40]: This and the few lines below are the main changes on the acas side.  I added backwards compatability for the old syntax so out of date customer specific server functions would work the same.  This passes the query parameters directly to roo.

File: modules/Components/src/server/routes/CodeTableServiceRoutes.coffee [hunk 5]
```diff
@@ -37,31 +37,33 @@ exports.getAllCodeTableValues = (req, resp) ->
...
+			qs = query
 		request = require 'request'
 		request(
 			method: 'GET'
 			url: baseurl
+			qs: qs
 			json: true
 		, (error, response, json) =>
 			if !error && response.statusCode == 200
-				if json.length > 0 and !json[0].displayOrder? and json[0].name?
```
Comment (brianbolt) [inline modules/Components/src/server/routes/CodeTableServiceRoutes.coffee:58]: I don't know why this was added. Roo always did this same sorting before passing the response and it was messing up my "best match" sort when searching by labelTextSerch so I removed it.

File: modules/ServerAPI/src/server/CustomerSpecificServerFunctions.coffee [hunk 6]
```diff
@@ -409,7 +409,7 @@ exports.getAllAuthors = (opts, callback) ->
 		# This is was added for the purpose of allowing additional non-authors to show up in picklists throughout ACAS and Creg
 		if opts.additionalCodeType? and opts.additionalCodeKind?
 			codeTableServiceRoutes = require "#{ACAS_HOME}/routes/CodeTableServiceRoutes.js"
-			codeTableServiceRoutes.getCodeTableValuesInternal opts.additionalCodeType, opts.additionalCodeKind, (codes) ->
+			codeTableServiceRoutes.getCodeTableValuesInternal opts.additionalCodeType, opts.additionalCodeKind, null, (codes) ->
```
Comment (brianbolt) [inline modules/ServerAPI/src/server/CustomerSpecificServerFunctions.coffee:412]: Everything below here is just to match the new codeTableValuesInternal syntax

File: modules/Components/src/client/PickList.coffee (hunk 2 above)
Comment (bffrost) [reply modules/Components/src/client/PickList.coffee:364]: Indeed this is kinda hacky but it's reasonable. You're programmatically getting the base path of the URL, so this looks reliable.

File: modules/CodeTablesAdmin/src/server/routes/CodeTablesAdminRoutes.coffee [hunk 7]
```diff
@@ -105,15 +105,7 @@ exports.searchCodeTablesEntities = (req, resp) ->
...
-	searchTerm = req.params.searchTerm.toLowerCase().trim()
-	codeTableServiceRoutes.getCodeTableValuesInternal req.params.codeType, req.params.codeKind, (results) ->
-		searchResults = []
-		if searchTerm == "*" || searchTerm == ""
-			searchResults = results
-		else
-			for r in results
-				if (r.code? && r.code.toLowerCase().indexOf(searchTerm) >= 0) || (r.comments? && r.comments.toLowerCase().indexOf(searchTerm) >= 0) || (r.description? && r.description.toLowerCase().indexOf(searchTerm) >= 0) || (r.name? && r.name.toLowerCase().indexOf(searchTerm) >= 0) 
-					searchResults.push(r)
+	codeTableServiceRoutes.getCodeTableValuesInternal req.params.codeType, req.params.codeKind, searchTerm, (results) ->
```
[the remark this answers is not in the document]
Comment (bffrost) [reply modules/CodeTablesAdmin/src/server/routes/CodeTablesAdminRoutes.coffee:108]: Ah, I see. This was an older node-server implementation of searching codeTables that we think was unused, but you've now ported it to use the database-based full text search.
I'm in favor of this change, and makes sense that we can tolerate the slight search behavior change given we don't think `searchCodeTablesEntities` is actually being used anywhere.

**Review by bffrost — APPROVED** [review]
LGTM. I appreciate the thought you put into backward compatibility and some of the minor cleanup you've done along the way.

File: modules/Components/src/client/PickList.coffee (hunk 2 above)
Comment (brianbolt) [reply modules/Components/src/client/PickList.coffee:364]: As a side note, this might be a way of fixing our relative vs full path urls problem with ACAS sometime int he future:
https://stackoverflow.com/questions/16553026/prepend-absolute-url-path-in-backbone-sync-for-an-api-on-another-server
